Added usecase process graphs

Pratichhya opened the pull request.

Pratichhya commented:
added process graphs and use case notebooks

Pratichhya commented:
added aoi

Pratichhya commented:
AOI repo updated in each file along with term correction.

m-mohr reviewed (changes requested):
Thank you for the submission. This needs a Readme tgat explains the use case and what the files are doing /used for. Also, this should be moved to a separate subfolder. Thanks.

Pratichhya commented:
Added a readme file

Pratichhya commented:
updated readme

jdries commented:
General remarks from review:
* for the process graph json files, can we have parameterized versions under the 'processes' directory?
* Try to keep the original table/layout in python/readme.md

Pratichhya commented:
updated readme

Pratichhya commented:
Parameterised processes graph and readme updated.

m-mohr commented on python/README.md:
```diff
@@ -1,15 +1,25 @@
 
-* Please provide each contribution in a separate folder.
\ No newline at end of file
+* OpenEO/EOplaza users can also send a pull request in this repository as either a python file/notebook for their usecase. Showcasing their usecases are highly encouraged.
```
```suggestion
* OpenEO users can send a pull request to this repository as either a Python file or a Jupyter notebook for their usecase. Submissions should be in separate folders. Showcasing their usecases are highly encouraged.
```

m-mohr commented on python/process_graphs/burntmapping_chunks.json:
```diff
@@ -0,0 +1,156 @@
+{
```
This is a not very useful process to have as an example here as it runs only on a single back-end and doesn't use any pre-defined processes from openEO.

jdries commented on python/process_graphs/biomass_basic.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "process_graph": {
+    "BIOMASS1": {
+      "process_id": "BIOMASS",
```
Indeed, this is more for eoplaza. @Pratichhya  Can you remove this one?

Pratichhya commented:
@m-mohr metadata has been updated in the process graph. Please let me know if it further needs to be improvised

jdries commented:
@m-mohr  would it be fine to merge this now? I have some potential changes to the notebooks, but would be easier if I can do that directly on this repo.

m-mohr commented:
I have another round of change requests, sorry:

- Could you please put every use case in a separate folder so that users can grab all files they need for a use case just by copying the folder? So biomass_basic notebook and AOI geojson in a biomass_basic folder etc?
- Please add links to the folder to the README (in the title column)
- I'll go through the other files now and add separate comments..

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+{
+  "namespace": "user",
+  "description": "Classical Normalized Burnt Ratio(NBR) difference is performed using VITO backend on a chunk polygon.",
+  "id": "NBR_chunk_polygons",
```
process ID should be the same as the filename

Pratichhya replied:
updated

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+{
+  "namespace": "user",
```
Please remove the namespace

Pratichhya replied:
removed

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+  ],
+  "parameters": [
+    {
+      "description": "Left-closed temporal interval, i.e. an array with exactly two elements:The first element is the start of the temporal interval. The specified instance in time is included in the interval.The second element is the end of the temporal interval. The specified instance in time is excluded from the interval.The specified temporal strings follow RFC 3339. Also supports open interv ...
```
```suggestion
      "description": "Left-closed temporal interval, i.e. an array with exactly two elements: The first element is the start of the temporal interval. The specified instance in time is included in the interval. The second element is the end of the temporal interval. The specified instance in time is excluded from the interval. The specified temporal strings follow RFC 3339. Also supports open intervals by setting one of the boundaries to null, but never both.",
```

Generally, please use spaces between dots and the next word.

Pratichhya replied:
Oops. Sorry about those. The typos are corrected now.

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+          "B08",
+          "B12"
+        ],
+        "id": "SENTINEL2_L1C_SENTINELHUB",
```
Hardcoded collection IDs (and band names) make this not easily re-usable.

Pratichhya replied:
Sorry I am not sure how I should address these.

m-mohr replied:
Depends on the use case, ideally you'd have parameters for them and via context you could even pass them into the UDFs, but AFAIK that's not supported by VITO right now. 
So let's not consider UDFs for now, but everything else could simply use process parameters (i.e. add a parameter to `parameters` and then use `{"from_parameter": "..."}`.

jdries replied:
For collection names, I would then rather go for using harmonized names, like 'SENTINEL2_L1C" rather than adding parameters. In many cases, process graphs are not so generic that you can apply them to a random collection.
If users do want to change this, I would say they can just modify it in the process graph?

m-mohr replied:
Okay, as this is a nice to have, I'd propose to change to:

```suggestion
        "id": "SENTINEL2_L1C",
```

Pratichhya replied:
updated

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+      "process_id": "reduce_dimension"
+    }
+  },
+  "public": true
```
Please also remove this

Pratichhya replied:
removed

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+{
+  "namespace": "user",
+  "description": "Classical Normalized Burnt Ratio(NBR) difference is performed using VITO backend on a chunk polygon.",
```
I think community examples should not be back-end specific.

Pratichhya replied:
corrected

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+      }
+    },
+    {
+      "description": "Spatial extent for area of interst to calculate NBR, specified as GEOJSON.",
```
```suggestion
      "description": "Spatial extent for area of interest to calculate NBR, specified as GEOJSON.",
```

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+  "namespace": "user",
+  "description": "Classical Normalized Burnt Ratio(NBR) difference is performed using VITO backend on a chunk polygon.",
+  "id": "NBR_chunk_polygons",
+  "summary": "Burnt area mapping using chunk_polygon on UDF",
```
```suggestion
  "summary": "Burnt area mapping",
```

Technical details better go into the description.

m-mohr commented on processes/burntmapping_chunks.json:
```diff
@@ -0,0 +1,196 @@
+  "description": "Classical Normalized Burnt Ratio(NBR) difference is performed using VITO backend on a chunk polygon.",
+  "id": "NBR_chunk_polygons",
+  "summary": "Burnt area mapping using chunk_polygon on UDF",
+  "returns": "Return an RGB image as Geotiff",
```
`returns` must be an object, not a string

m-mohr commented on processes/flood_ndwi.json:
```diff
@@ -0,0 +1,123 @@
+{
+  "namespace": "user",
+  "description": "Performed comparative study between pre and post image for Colong during 2021 flood. A simple technique to subtract pre and post image is done to know the change in water content due to flood in that region.",
```
```suggestion
  "description": "Performed comparative study between pre and post image for Colone during 2021 flood. A simple technique to subtract pre and post image is done to know the change in water content due to flood in that region.",
```

m-mohr commented on processes/flood_sar_udf.json:
```diff
@@ -0,0 +1,200 @@
+      }
+    },
+    {
+      "description": "Spatial extent for area of interst to calculate NBR, specified as GEOJSON.",
```
```suggestion
      "description": "Spatial extent for area of interest to calculate NBR, specified as GEOJSON.",
```

Pratichhya replied:
corrected

m-mohr commented on processes/rescale_chunks.json:
```diff
@@ -0,0 +1,99 @@
+      }
+    },
+    {
+      "description": "Spatial extent for area of interst to calculate NBR, specified as GEOJSON.",
```
```suggestion
      "description": "Spatial extent for area of interest to calculate NBR, specified as GEOJSON.",
```

Pratichhya replied:
corrected

m-mohr commented on processes/flood_ndwi.json:
```diff
@@ -0,0 +1,123 @@
+      }
+    },
+    {
+      "description": "Spatial extent for area of interst to calculate NBR, specified as GEOJSON.",
```
```suggestion
      "description": "Spatial extent for area of interest to calculate NBR, specified as GEOJSON.",
```

Pratichhya replied:
corrected

Pratichhya commented:
- Kept each usecase in a separate folder with their AOI.
- Updated the README file

It's ok you can let me know if additional changes needs to be made

m-mohr commented on processes/flood_sar_udf.json:
```diff
@@ -0,0 +1,200 @@
+{
+  "namespace": "user",
+  "description": "Flood extent can be determined using a change detection approach on Sentinel-1 data. In this notebook we have tried in adopting (UN SPIDER's) recommendation practice for computing flood extent by implementing openeo process.Function) UDF.",
```
```suggestion
  "description": "Flood extent can be determined using a change detection approach on Sentinel-1 data. In this process, we have tried adopting UN SPIDER's recommended practice for computing flood extents by implementing an openEO UDF.",
```

Is that what you meant?

m-mohr commented on processes/rescale_chunks.json:
```diff
@@ -0,0 +1,99 @@
+{
+  "namespace": "user",
+  "description": "It is a simple process to rescale Sentinel 2 RGB image along with the use of chunk_polygon  apply with a (User Defined Function) UDF.",
```
```suggestion
  "description": "It is a simple process to rescale Sentinel 2 RGB image along with the use of ``chunk_polygon()`` apply with a (User Defined Function) UDF.",
```

I don't understand this sentence. Are there words missing or so?

And you can link to other processes with the markup given above. :-)

Pratichhya replied:
modified

m-mohr reviewed (changes requested):
I've reviewed just one process completely, but most comments apply to all processes, so please update them all accordingly although there are no specific comments available for each of them.

m-mohr commented on python/FloodSAR/flood_sar_udf.ipynb:
```diff
@@ -0,0 +1,224 @@
+    "# load before flash flood params\n",
+    "before_date = [\"2021-05-12\",\"2021-05-12\"]\n",
+    "after_date = [\"2021-06-18\", \"2021-06-18\"]\n",
+    "spatial_param = read_json(\"../aoi/cologne_all.geojson\")\n"
```
After moving to separate folders per use-case the paths in the Jupyter Notebooks are not correct anymore, right?

Pratichhya replied:
updated

m-mohr commented on python/README.md:
```diff
@@ -1,15 +1,23 @@
 
 Environments: `Python` (plain Python code), `Jupyter` (e.g. Notebooks)
 
+**Furthermore in usecase_notebooks repository:**
```
This needs to be removed.

Pratichhya replied:
updated

m-mohr commented on python/README.md:
```diff
@@ -1,15 +1,23 @@
 | Title | Environment | Description |
 | ----- | ----------- | ----------- |
-| ...   | `Python`    | ...         |
+| BurntMapping   | `Jupyter`    | Classical Normalized Burnt Ratio(NBR) difference performed using VITO backend on a chunk polygon. The method followed in this notebook to compute DNBR is inspired from [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping). |
```
Please add links to the individual folders:

```suggestion
| [BurntMapping](./BurntMapping/)   | `Jupyter`    | Classical Normalized Burnt Ratio(NBR) difference performed using VITO backend on a chunk polygon. The method followed in this notebook to compute DNBR is inspired from [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping). |
```

Pratichhya replied:
update

m-mohr commented on python/README.md:
```diff
@@ -1,15 +1,23 @@
 | ----- | ----------- | ----------- |
-| ...   | `Python`    | ...         |
+| BurntMapping   | `Jupyter`    | Classical Normalized Burnt Ratio(NBR) difference performed using VITO backend on a chunk polygon. The method followed in this notebook to compute DNBR is inspired from [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping). |
+| FloodNDWI   | `Jupyter`    | Comparative study between pre and post image for Colong during 2021 flood using NDWI. [Refernce](https://labo.obs-mip.fr/multitemp/the-ndwi-applied-to-the-recent-flooding-in-the-central-us/) |
```
```suggestion
| FloodNDWI   | `Jupyter`    | Comparative study between pre and post image for Colone during 2021 flood using NDWI. [Refernce](https://labo.obs-mip.fr/multitemp/the-ndwi-applied-to-the-recent-flooding-in-the-central-us/) |
```

Pratichhya replied:
updated

m-mohr commented on python/README.md:
```diff
@@ -1,15 +1,23 @@
-| ...   | `Python`    | ...         |
+| BurntMapping   | `Jupyter`    | Classical Normalized Burnt Ratio(NBR) difference performed using VITO backend on a chunk polygon. The method followed in this notebook to compute DNBR is inspired from [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping). |
+| FloodNDWI   | `Jupyter`    | Comparative study between pre and post image for Colong during 2021 flood using NDWI. [Refernce](https://labo.obs-mip.fr/multitemp/the-ndwi-applied-to-the-recent-flooding-in-the-central-us/) |
+| FloodSAR   | `Jupyter`    | Flood extent detection following [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping) using openeo process.|
```
```suggestion
| [FloodSAR](./FloodSAR/)   | `Jupyter`    | Flood extent detection following [UN SPIDER's recommended practices](https://www.un-spider.org/advisory-support/recommended-practices/recommended-practice-google-earth-engine-flood-mapping). |
```

m-mohr commented:
Quick note @Pratichhya: I've updated your branch to include the new CI that checks the processes for validity and also lints them to follow best practices. This should help you to check your processes, it requires nodejs being installed. After installing nodejs, you can run `npm install` (once to install dependencies) and `npm test` to check your contributon for compliance.

Pratichhya reviewed (commented):
updated

m-mohr commented:
Happy to merge the notebook, but the processes are not passing the tests yet. Please review the issues and fix them. Let me know if you need help with running the tests.

Pratichhya commented:
> Happy to merge the notebook, but the processes are not passing the tests yet. Please review the issues and fix them. Let me know if you need help with running the tests.

@m-mohr I guess I now need a bit of help. Indentations are corrected, and including a specific format for metadata is improvised. Yet, there are a few spelling issues, which seems fine to me. How should I address them? Also, please let me know if  I missed other points.

Pratichhya commented:
> Words that are clearly correct but are not covered by the dictionary of the linter can be added to this ignore file: `https://github.com/Open-EO/openeo-community-examples/blob/main/processes/tests/.words`. One word per line.

@m-mohr I tried in including those words in this ignore file but still had the same issue. Also, there are a few other issues related to the summary, and also I am not sure, but it raises an issue that the Category is not defined.

m-mohr commented on processes/rescale_chunks.json:
```diff
@@ -2,7 +2,9 @@
     "id": "rescale_chunks",
     "Summary": "Rescaling of RGB within chunk of polygons.",
     "description": "A simple process for rescaling Sentinel 2 RGB images within polygon chunks that also showcases how to use ``chunk_polygon() with a (User Defined Function) UDF.",
```
```suggestion
    "description": "A simple process for rescaling Sentinel 2 RGB images within polygon chunks that also showcases how to use ``chunk_polygon()`` with a (User Defined Function) UDF.",
```

m-mohr replied:
Although the process has recently been renamed in the spec to apply_polygon.

m-mohr commented on processes/tests/.words:
```diff
@@ -1,5 +1,5 @@
-Sentinel-1
-chunk_polygon
\ No newline at end of file
+*OpenEO-based
```
It should work without any special characters at the beginning. If it doesn't pick them up there might be another issue in the tooling.

Pratichhya replied:
I tried them without any special characters or with . and this was with *

m-mohr replied:
It was a casing issue... openEO != OpenEO ;-)

m-mohr commented on processes/rescale_chunks.json:
```diff
@@ -2,7 +2,9 @@
     "Summary": "Rescaling of RGB within chunk of polygons.",
     "description": "A simple process for rescaling Sentinel 2 RGB images within polygon chunks that also showcases how to use ``chunk_polygon() with a (User Defined Function) UDF.",
-    "Categories": ["arrays"],
+    "Categories": [
```
```suggestion
    "categories": [
```

Generally lower-case.

Pratichhya replied:
i think i tried with small initially, but I can give a try again. Similar was with Summary

Pratichhya replied:
Sorry, it does work for categories. But not for summary

m-mohr replied:
For summaries it may complain about the dots at the end or the length. 

Otherwise, I can give it a try later.

m-mohr replied:
Indeed, it was either casing issues or the dots at the end of the summaries.
The linter error messages were not very helpful, unfortunately. Something to improve in the linter code... :-)

Pratichhya commented:
> @Pratichhya I've fixed the CI issues for you, it now validates. I had a final look at the processes, but the return values seem to not completely match the actual behavior. For example in `anomaly_detection` the return value is described as boolean value, but the last process is a UDF so I assume it returns more than just a boolean value (a data cube instead?) Please check the return values again. Otherwise, it seems to look good.

@m-mohr sure I will correct those. Actually, initially, I wrote what is returned as the final output which gave an error so I replaced it with this common format. Also, is there anything that I could follow to be specific with these graphs?

m-mohr commented:
@Pratichhya I can't follow. Can you give an example? If your last node is merge_cubes though it likely makes sense to look at merge_cubes and copy the schema from that process. The description is then just a text that describes more in detail what a user can expect, e.g. a data cube with the dimension a,b,c and the values are ...

Oh and I just saw that you still need to add the processes to the table in processes/README.md, please.

Pratichhya commented:
> 

Thank you. Sure will do that

Pratichhya commented:
@m-mohr can you please help me with the return type?

m-mohr commented on processes/anomaly_detection.json:
```diff
@@ -71,7 +71,7 @@
         "description": "A file with the CropSAR fAPAR values of the individual fields and the regional average fAPAR value for the specified time period",
         "schema": {
-            "type": "data cube"
+            "type": "DataCube"
```
type: object
subtype: raster-cube

Pratichhya replied:
thank you. Updated

m-mohr replied:
Hmm... the return values don't make sense to me as they are. The description of burntMapping_chunks says it returns a GeoTiff, but the schema says it returns a raster data cube. The raster data cube seems correct due to the last node being merge_cubes. So this needs a description that describes how the data cube looks like. What dimensions are returned? What about the values? ...

Pratichhya commented on processes/anomaly_detection.json:
```diff
@@ -0,0 +1,1508 @@
+                        "arguments": {
+                            "data": {
+                                "from_node": "resamplecubespatial1",
+                                "node": {
```
@m-mohr For this openeo process, they perform some mandatory calculations before using it as a final process. I guess that is represented inside a node. I confirmed with my colleague that it should be one of the steps for this process, and we are not sure if removing it will give the desired outcome.

m-mohr replied:
I really meant the JSON key "node" here, not an openEO node in the process graoh.  The key "node" is undefined alongside a "from_node" and should be removed.

Pratichhya replied:
@m-mohr, the process graph is updated.

m-mohr replied:
Great, thanks.

m-mohr commented on processes/README.md:
```diff
@@ -8,6 +8,11 @@ This folder contains user-defined processes that can be run by any openEO client
 | -- | ---------- | ------- |
 | [array_contains_nodata](array_contains_nodata.json) | arrays | Check for no-data values in an array |
 | [array_find_nodata](array_find_nodata.json)         | arrays | Find no-data values in an array |
+| [anomaly_detection](anomaly_detection.json)         | arrays | Regional Benchmarking using CropSAR |
```
The categories seem wrong. Can you update these to the categories used in the process files?

Pratichhya replied:
I updated it to DataArray, but I not sure that if is that was what is to be specified as categories

m-mohr commented on python/README.md:
```diff
@@ -6,9 +6,17 @@ This folder contains examples and code snippets for the openEO Python client.
 | Title | Environment | Description |
 | ----- | ----------- | ----------- |
-| ...   | `Python`    | ...         |
+| [Anomaly_Detection](./RescaleChunks/)   | `Python`  | Check the crop growth on your field and compare it with similar fields in the region.|
```
These are all Python Notebooks as far as I can see, so the Environment should be set to `Jupyter` instead of `Python`

Pratichhya commented:
> Looks good, just needs the two changes in the README files and then we can merge this.

m-mohr commented:
Where does the "DataArray" come from? That seems Python(?) specific? Better use [existing categories](https://processes.openeo.org) from openEO and add some specific categories that reflect the usecase, e.g. a selection of "cubes", "math", "sar", "optical", "udf", "vegetation indices" and/or "math > indices" depending on the use case.

Pratichhya commented:
> Where does the "DataArray" come from? That seems Python(?) specific? Better use [existing categories](https://processes.openeo.org) from openEO and add some specific categories that reflect the usecase, e.g. a selection of "cubes", "math", "sar", "optical", "udf", "vegetation indices" and/or "math > indices" depending on the use case.

@m-mohr, sorry, I was not sure they were OpenEO categories specific, I have updated them. Hope it looks fine now.

m-mohr commented:
Thanks, merged. Feel free to open new PRs to update the files in the future.

Pratichhya commented:
> Thanks, merged. Feel free to open new PRs to update the files in the future.

Thank you :)
